### PR TITLE
Support for saving and loading pytorch compatible state dictionaries

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**: When retrieving the ```state_dict``` from ```ORTModule```, the key names had the ```_flattened_output_module._base_module``` prefix for all states (because of the specialized module inside ```ORTModule``` called ```self._flattened_output_module._base_module```). As a result, the loading of the state dictionary from a pytorch model also failed (since ```ORTModule``` expected the prefix to be there for all states).

This pull request overrides the ```state_dict()``` and the ```load_state_dict()``` methods and calls the same methods on the underlying specialized ```torch.nn.Module```.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -503,3 +503,17 @@ def _get_inference_graph(self, *inputs, **kwargs):
...
+        # do not contain the _flattened_output_module._base_module prefix
+        return self._flattened_output_module._base_module.state_dict(
+            destination=destination, prefix=prefix, keep_vars=keep_vars)
+
+    def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
+                        strict: bool = True):
+        # Override the load_state_dict() method so that the loaded state dict
+        # key names does not need to contain the _flattened_output_module._base_module prefix
+        return self._flattened_output_module._base_module.load_state_dict(
+            state_dict, strict=strict)
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:518]: Thanks for making this change! Are there any other methods that we should override in a similar way, for consistency?

e.g. I could imagine that any method dealing in parameter *names* should be delegated to the base module. So maybe `named_parameters()`, `named_buffers()`, others...?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:518]: We can start with:
- ```state_dict()```
- ```load_state_dict()```
- ```parameters()```
- ```named_parameters()```
- ```buffers()```
- ```named_buffers()```
for now. And can add support as required.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -503,3 +504,40 @@ def _get_inference_graph(self, *inputs, **kwargs):
...
+
+    def register_buffer(self, name: str, tensor: Optional[torch.Tensor], persistent: bool = True) -> None:
+        self._flattened_output_module._base_module.register_buffer(name, tensor, persistent=persistent)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:522]: could you add the same comments for all other overrides? Maybe summarize to fit in a single line

`'''Overload original method to delegate execution to the base module'''`

Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:522]:
```suggestion
        '''Overload original method to delegate execution to the base module'''

        self._flattened_output_module._base_module.register_buffer(name, tensor, persistent=persistent)
```

File: orttraining/orttraining/python/training/ortmodule.py (hunk 1 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:519]: I also added a few more methods:
- ```register_buffer()```
- ```register_parameter()```
- ```get_buffer()```
- ```get_parameter()```

**Review by thiagocrepaldi — APPROVED** [review]

**Review by mrry — APPROVED** [review]

**Pull request merged**